File: Intersection_3Array/intersection.cs
using System.Collections.Generic;

public class Solution
{
    public IList<int> ArraysIntersection(int[] arr1, int[] arr2, int[] arr3)
    {
        List<int> result = new List<int>();

        for (int i = 0; i < arr1.Length; i++)
        {
            if (binarySearch(arr1[i], arr2))
            {
                result.Add(arr1[i]);
            }
        }

        for (int i = result.Count - 1; i >= 0; i--)
        {
            if (!binarySearch(result[i], arr3))
            {
                result.RemoveAt(i);
            }
        }

        return result;
    }

    bool binarySearch(int val, int[] target)
    {
        int start = 0,
            end = target.Length - 1;

        while (start <= end)
        {
            int middle = (int)((start + end) / 2);

            if (target[middle] > val)
            {
                end = middle - 1;
            }
            else if (target[middle] < val)
            {
                start = middle + 1;
            }
            else
            {
                return true;
            }
        }

        return false;
    }
}

File: Intersection_3Array/intersection.py
from typing import List


class Solution:
    def arraysIntersection(
        self, arr1: List[int], arr2: List[int], arr3: List[int]
    ) -> List[int]:
        intersection: List[int] = []

        for num in arr1:
            if binary_search(num, arr2):
                intersection.append(num)

        index = len(intersection) - 1

        while index >= 0:
            if not binary_search(intersection[index], arr3):
                intersection.remove(intersection[index])

            index -= 1

        return intersection


def binary_search(val: int, target: List[int]) -> bool:
    start, end = 0, len(target) - 1

    while start <= end:
        middle = (start + end) // 2
        if target[middle] < val:
            start = middle + 1
        elif target[middle] > val:
            end = middle - 1
        else:
            return True

    return False
